<!DOCTYPE html>
<html>
	<head>
		<title>ProteoSign</title>
		<link rel="stylesheet" type="text/css" href="css/style.css">
		<link rel="stylesheet" type="text/css" href="css/jquery-ui.min.css">
		<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
		<meta name="keywords" content="ProteoSign, mass spectrometry, proteomics, data, statistics, statistical analysis, bioinformatics, limma, comparative proteomics, silac, labeling, differential expression">
		<meta charset="UTF-8">
		<link rel="shortcut icon" href="favicon.ico" type="image/x-icon">
	</head>
	<body>
		<div class="main_div">
			<div class="main_div main_header">
				<span class="logo">ProteoSign</span><br>
				<span class="logo" style="font-size: 100%; left: 12%">comparative proteomics</span>
				<a href="help.html" target="_blank"><i class="fa fa-question-circle helplink" aria-hidden="true"></i></a><a class="helplink" href="help.html" target="_blank"> HELP</a>
			</div>
			<div class="main_div main_nav">
				<p class="scrolling_p">ProteoSign Change Log</p>
			</div>
			<div class="main_div main_section" id="s1div">
				<ul>
					<li>Added support for <strong>Replication Multiplexing</strong>! Now you can use <strong>ProteoSign</strong> for experiments where biological, technical replicates or conditions are represented either as different tags or as different MS runs. Visit our <a href = "help.html#RM">help page</a> for more information!
				</ul>
				<ul>
					<li>We integrated <strong>expression enrichment analysis</strong> to <i>ProteoSign's</i> pipeline! Choose the target organism of your experiment in <i>Step 3 </i> and get useful information concerning your differentially expressed proteins' function - pathways etc.!
				</ul>
				<ul>
					<li><i>ProteoSign</i> is now easy to install on a local server! Check our <a href = "http://bioinformatics.med.uoc.gr/ProteoSign/documentation/index.html">documentation</a> for more information!
				</ul>
				<ul>
					<li><strong>Venn diagrams</strong> are very useful to evaluate the reproducibility of your experiments! <i>ProteoSign</i> now generates Venn diagrams for proteins' quantification consistency across replicates and conditions. Our <a href = "http://bioinformatics.med.uoc.gr/ProteoSign/documentation/index.html">documentation</a> will guide you in getting the most out of this new feature!
				</ul>
			</div>
<div class="main_div main_footer">
				<p>
					<em> </em><br>
				<p>
				<a href="http://bioinformatics.med.uoc.gr/ProteoSign/"><span id="proteosignversion" style='font-family: "Courier New"; display: block; float: left'>Back to ProteoSign</span></a>
				</span></div>
		<div>
		<script src="http://code.jquery.com/jquery-2.1.0.js"></script>
		<script src="js/jquery.dataTables.min.js"></script>
 		<script src="js/jquery-ui.min.js"></script>
		<script src="js/jquery.ui-contextmenu.min.js"></script>
		<script src="js/dataTables.select.min.js"></script>
		<script src="js/script.js"></script>
</body>
</html>
