<!DOCTYPE html>
<html>
	<head>
		<title>ProteoSign</title>
		<link rel="stylesheet" type="text/css" href="css/style.css">
		<meta name="keywords" content="ProteoSign, mass spectrometry, proteomics, data, statistics, statistical analysis, bioinformatics, limma, comparative proteomics, silac, labeling, differential expression">
		<meta charset="UTF-8">
	</head>
	<body>
		<div class="main_div">
			<div class="main_div main_header">
				<span class="logo">ProteoSign</span><br>
				<span class="logo" style="font-size: 100%; left: 12%">comparative proteomics</span>
			</div>
			<div class="main_div main_nav">
				<p id="scrollingtext" class="scrolling_p"></p>
			</div>
			<div class="main_div main_section" id="s1div">
				<p><strong>ProteoSign</strong> is an online service for protein <strong>differential expression (or abundance) analysis</strong> designed with the end-proteomics user in mind.</p>
				<p>Borrowing from the experience of the microarray field, ProteoSign software utilizes the well-established <a href="http://bioinf.wehi.edu.au/limma/" target="_blank"><strong>Linear Models For Microarray Data (LIMMA)</strong></a> methodology for assessing statistical significance in protein abundance differences between two or more proteome states.</p>
				<p>ProteoSign aims to fully automate the process of statistically evaluating differential expression in mass spectrometry-based bottom-up (shotgun) quantitative proteomics data by requiring <strong>minimal user interaction time</strong> and generating <strong>publication-quality data plots.</strong></p>
				<br><hr class="style-one"><br>
				<p style="color: #000000">Click <em>Start</em> to begin your analysis.</p>
				<br>
				<p style="font-size: 0.9em"><strong>Important note: In the case of an unexpected error, please help us to improve this software by <a id='contact' onclick='$("#contact").attr("href","mailto:msdiffexp@gmail.com?Subject=Session%20"+sessionid); return true;' target='_blank'><u>contacting</u></a> us.</strong></p>
				<div class="buttonbar">
					<button type="button" class="main" id="s1btnf">Start</button>
				</div>
			</div>
			<div class="main_div main_section hidden" id="s2div">
				<h2>Step 1</h2>
				<p style="color: #000000">Please read below first. When ready, upload your dataset and click <em>Next</em>. Alternatively, you could <a href="#" onclick="onChooseFromTestDatasets();"><u>choose</u></a> from available test datasets.</p>
				<p>
				<dl style="font-family: 'Courier New'; font-size: 75%; padding-left: 5%;">
					<dt>Input file support:</dt>
					<dd>&#8212 <a href="http://www.google.com/search?q=Thermo%20Proteome%20Discoverer&btnI" target="_blank">Proteome Discoverer&#8482</a> (PD) v1.4+ PSM file.<span class="tooltip" style="position: relative">?<div style="left: -25em; min-width: 50em;"><img class="callout" src="images/callout_black.gif" style="margin-left: 24.3em"/>The <em>peptide spectrum match</em> (PSM) file, comprising the necessary information required for protein quantitation, has to be produced manually through the PD desktop application (see the <em>Export&#x2192;To Text</em> option under the <em>File</em> menu).<p>The PSM file must be created from within a PD report which combines the information from all relevant MS analyses (i.e. replicates, conditions etc). Such a PD report is sometimes required to be created based on separate reports and is referred to as a multi-consensus report.</p><p>Before exporting the information, one <strong>must</strong> first disable peptide grouping and second edit the report's <em>Quantification&nbsp;Method</em> by configuring the <em>Ratio&nbsp;Calculation</em> parameters as shown below. The following screen-shots show where the aforementioned changes can be made from within the PD application.</p><figure style="float: left; width: 49%; margin: 0 auto;"><img src="images/pdpepgroup.png" style="width: 100%;"/><figcaption style="text-align: center;">Disabling peptide grouping.</figcaption></figure><figure style="float: right; width: 49%; margin: 0 auto;"><img src="images/pdquantmeth.png" style="width: 100%;"/><figcaption style="text-align: center;">Editing ratio calculation parameters.</figcaption></figure></div></span></dd>
					<dd>&#8212 <a href="http://www.google.com/search?q=Max%Planck%MaxQuant&btnI" target="_blank">MaxQuant</a> (MQ) v1.3.0.5+ 'proteinGroups' and 'evidence' files.</dd>
					<dt><br>Experiment support:</dt>
					<dd>&#8212 Data processed with PD: Label-free, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=15385600" target="_blank">iTRAQ</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=12713048" target="_blank">TMT</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=12118079" target="_blank">SILAC</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=19053139" target="_blank">pulsed SILAC</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=19300442" target="_blank">dimethyl</a>.</dd>
					<dd>&#8212 Data processed with MQ: <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=12118079" target="_blank">SILAC</a>, <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=19053139" target="_blank">pulsed SILAC</a>.</dd>
					<dt><br>Assumptions made by the software:</dt>
					<dd>&#8212 The alphabetical order of raw-data file names must correspond to the following sample order structure, for N technical replicates and M biological replicates: technical replicate #1 of biological-replicate #1 (b1t1), technical replicate #2 of biological-replicate #1 (b1t2), ..., b1tN, b2t1, ..., b2tN, bMt1, ..., bMtN. If the samples where not acquired in such order, then the raw-data file names must be renamed accordingly before pre-processing with PD or MQ.</dd>
				</dl>
				</p>
				<!-- Actual file chooser button, hidden (it can't be avoided) -->
				<input type="file" id="__s2btnupld" style="display: none" multiple/>
				<p>
					<button class="main upload" type="file" id="s2btnupld"/>Upload file(s)</button>
				</p>
				<!-- File uploaders will be rendered below when user selects file(s) -->
				<div class="sectionForm" id="s2uluploaders" style="height: 20%">
					<table style="border-spacing: 15px 5px;">
					</table>
				</div>
				<div class="buttonbar">
					<button type="button" class="main" id="s2btnb">Back</button>
					<button type="button" class="main" id="s2btnf" disabled>Next</button>
				</div>
			</div>
			
			<div class="main_div main_section hidden" id="s22div">
				<h2>Step 2</h2>
				<p style="color: #000000">Please define the experimental design by assigning to each raw file the appropriate location coordinate in the experimental structure. 
					<p style="font-size: 14px;">A coordinate specifies the biological replicate, technical replicate and fraction a raw file belongs to. Hold down the <em>Ctrl</em> key and click to select individual items. Hold down the <em>Shift</em> and click to select a range of items. To assign technical replicates, specify the biological replicate and leave the technical replicate ordinate field blank. To assign fractions, specify both the biological and the technical replicate.</p>
				</p>
				<div style="text-align: center; height: 50%; overflow-y:scroll;">
					<table class="rawfiles_tbl" id="rawfiles_tbl_allfiles" style="float: left;">
						<tbody>
							<tr>
								<th>Raw files</th>
							</tr>
						</tbody>
					</table>
					<table class="expstructcoord_tbl" id="expstructcoord" style="float: left;">
						<tr>
							<td colspan="2">Experimental structure coordinates</td>
						</tr>
						<tr>
							<td colspan="2"><input id="expstructcoord_biorep" type="text" placeholder="Bio. replicate #" onkeypress="inputChCheck(event,'^[0-9]+$',2)"/></td>
						</tr>
						<tr>
							<td colspan="2"><input id="expstructcoord_techrep" type="text" placeholder="Tech. replicate #" onkeypress="inputChCheck(event,'^[0-9]+$',2)"/></td>
						</tr>
						<tr>
							<td><button type="button" class="main inline" id="btnAssignExpStructCoord">Assign</button></td>
							<td><button type="button" class="main inline" id="btnResetExpStructCoord" disabled>Reset</button></td>
						</tr>					
					</table>
				</div>
				<div class="buttonbar">
					<button type="button" class="main" id="s22btnb">Back</button>
					<button type="button" class="main" id="s22btnf" disabled>Next</button>
				</div>
			</div>
			
			<div class="main_div main_section hidden" id="s3div">
				<h2>Step 3</h2>
				<p style="color: #000000">Define the experimental parameters and click <em>Submit</em>.</p>
				<div class="sectionForm" id="s3expparams">
					<table style="border-spacing: 5px 5px;">
						<tr>
							<td>&#8212 Quantitation by Proteome Discoverer&#8482</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" />Leave <em>blank</em> for MaxQuant</span></span></td>
							<td><input name="exppddata" type="checkbox"/></td>
							<td></td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>&#8212 Experiment ID/name</td>
							<td></td>
							<td><input data-required="true" name="expid" type="text" onkeypress="inputChCheck(event,'^(?!_)[a-zA-Z0-9_]+$',20)" placeholder="Character rules apply"/></td>
							<td></td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>&#8212 Time point</td>
							<td></td>
							<td><input data-required="true" name="exptpoint" type="text" onkeypress="inputChCheck(event,'^[0-9\\.]+[hmsdwy]?$',6)" placeholder="Character rules apply"/></td>
							<td></td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>&#8212 Nubmer of biological replicates</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" />The sum of biological and technical replicates must be at least <strong>3</strong>.</span></span></td>
							<td><input data-required="true" name="expbioreps" type="text" onkeypress="inputChCheck(event,'^[0-9]+$',2)" placeholder="Criteria apply"/></td>
							<td></td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>&#8212 Nubmer of technical replicates</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" />The sum of biological and technical replicates must be at least <strong>3</strong>.</span></span></td>
							<td><input data-required="true" name="exptechreps" type="text" onkeypress="inputChCheck(event,'^[0-9]+$',2)" placeholder="Criteria apply"/></td>
							<td></td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>&#8212 Species label #1</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" /><strong>Warning!</strong><br><u>The order of labels defined here matters</u>. Define your labels in the same order they were defined in <a id="quantsoftdetectedspan"></a>. If there exist unlabeled species, please define them in the <em>advanced parameters</em> section below.</span></span></td>
							<td><select data-required="true" id="explbl1name_" name="explbl1name" type="text" placeholder="Name"></select></td>
							<td><input data-required="true" id="explbl1definition" name="explbl1def" type="text" placeholder="Definition" readonly class='hidden'/></td>
							<td><button type="button" class="main inline" id="btnAddLabel">+</button></td>
							<td><button type="button" class="main inline" id="btnRemLabel" disabled>&#8211</button></td>
						</tr>
					</table>
					<p><u class="astyle" id="s3showhideadvparams">Show advanced parameters</u></p>
					<table id="s3advparams" class="hidden" style="border-spacing: 5px 5px;">
						<tr class="hidden">
							<td>&#8212 Protein (as opposed to peptide) quantitation?</td>
							<td></td>
							<td><input name="expprotquant" type="checkbox" checked/></td>
						</tr>
						<tr>
							<td>&#8212 Quantitation filtering?</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" /><strong>Warning!</strong><br><u>For labelled experiments only</u>. If enabled, two options will become available: <strong>a)</strong> singlets (peptides with no signal at the MS level in all but one labelled version) can be excluded from protein quantitation (this is referred to as <u>peptide-level filtering</u>).<br><strong>b)</strong> Proteins identified just by peptides with a certain label can be excluded from the analysis (this is referred to as <u>protein-level filtering</u>).</span></span></td>
							<td><input name="expquantfilt" type="checkbox"/></td>
						</tr>
						<tr class='hidden'>
							<td>&#8212 Filtering based on which label?</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" />The singlet label.</span></span></td>
							<td><select name="expquantfiltlbl"></select></td>
						</tr>
						<tr class='hidden'>
							<td>&#8212 Type of filtering: Peptide-level (as opposed to protein-level)?</td>
							<td><span class="tooltip">?<span><img class="callout" src="images/callout_black.gif" /><strong>Warning!</strong><br><u>Peptide-level filtering currently supported just for PD data</u>.</span></span></td>
							<td><input name="expquantfiltprot" type="checkbox"/></td>
						</tr>
						<tr>
							<td>&#8212 Non-labelled "background" species present?</td>
							<td></td>
							<td><input name="explbl00" type="checkbox"/></td>
						</tr>
						<tr class='hidden'>
							<td>&#8212 "Background" species</td>
							<td></td>
							<td><select id="explbl0name_" name="explbl0"></select></td>
						</tr>						
					</table>					
				</div>
				<div class="buttonbar">
					<button type="button" class="main" id="s3btnb">Back</button>
					<button type="button" class="main" id="s3btnf">Submit</button>
				</div>
			</div>
			<div class="main_div main_section hidden" id="s4div">
				<h2>Step 4</h2>
				<p id="results_p" style="color: #000000"></p>
				<br>
				<hr class="style-one">
				<br>
				<div id="server_feedback" class="image_array_div">
				</div>
				<div class="buttonbar">
					<button type="button" class="main" id="s4btnb">Back</button>
					<button type="button" class="main" id="s4btnf" disabled>Next</button>
				</div>
			</div>						
			<div class="main_div main_section hidden" id="s5div">
				<h2>End</h2>
				<p style="color: #000000">Please download your results. Click <em>Reset</em> to start again. Thank you for using ProteoSign.</p>
				<div id="s5dndres">
					<br>
					<hr class="style-one">
					<br>
					<div id="dndres" style="text-align: center"></div>
				</div>				
				<div class="buttonbar">
					<button type="button" class="main" id="s5btnb">Back</button>
					<button type="button" class="main" id="s5btnf">Reset</button>
				</div>
			</div>						
			<div class="main_div main_footer">
				<p>
					<em>Reference:</em><br>
					<a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=ProteoSign%3A+An+end-user+on-line+service+offering+automated+protein+differential+expression+analysis+given+pre-processed+quantitative+proteomics+datasets." target="_blank"><strong>Efstathiou G., Trudgian D. C., Thomas B., Pavlopoulos G., Papanikolaou N., Acuto O. and Iliopoulos I.<br>ProteoSign: An end-user on-line service offering automated protein differential expression analysis given pre-processed quantitative proteomics datasets.</strong><br></a>
				<p>
				<span id="proteosignversion" style='font-family: "Courier New"; display: block; float: left'></span>
				<span id="githubproject" style='font-family: "Courier New"; margin: 0 18%;'><a href="https://github.com/yorgodillo/ProteoSign" target="_blank">Project hosted at GitHub <img src="images/github_icon.png" height="30" style="position: relative; vertical-align: top; margin: -0.7em 0;"></img></a></span>
				<span style='font-family: "Courier New"; display: block; float: right'><a id='contact' onclick='$("#contact").attr("href","mailto:msdiffexp@gmail.com?Subject=Session%20"+sessionid); return true;' target='_blank'><u>Contact</u></a></span>
			</div>
		<div>

		<div class="expparamsDlg" id="s3expparamsDlgLabels">
			<div style="text-align: center">
				<span style="font-size: 80%">Select one or more (by holding the <em>Ctrl</em> key).</span>
				<select id="s3expparamsDlgLabelsSelection" style="min-width:90%" multiple>
				</select>
			</div>
			<div class="buttonbar" style="text-align: center; position: relative; width: auto; margin-top: 1em">
				<button type="button" class="main inline" style="width: 4.2em" id="dlglabelsBtnOK">OK</button>
				<button type="button" class="main inline" style="width: 4.2em; margin-left: 1em" id="dlglabelsBtnCancel">Cancel</button>
			</div>			
		</div>
		
		<div class="expparamsDlg" id="s1TestDatasets">
			<div style="text-align: center">
				<span style="font-size: 80%">Select a dataset.</span>
				<select id="s1TestDatasetsSelection" style="min-width:90%">
				</select>
			</div>
			<div class="buttonbar" style="text-align: center; position: relative; width: auto; margin-top: 1em">
				<button type="button" class="main inline" style="width: 4.2em" id="dlgTestDatasetsBtnOK">OK</button>
				<button type="button" class="main inline" style="width: 4.2em; margin-left: 1em" id="dlgTestDatasetsBtnCancel">Cancel</button>
			</div>			
		</div>		

		<script src="http://code.jquery.com/jquery-2.1.0.js"></script>
		<script src="js/script.js"></script>
	</body>
</html>
